README.md
# Twig + Markdown HTML Generator

![GitHub](https://img.shields.io/github/license/andi1984/back2roots)
![GitHub tag (latest by date)](https://img.shields.io/github/v/tag/andi1984/back2roots?label=Version)
![Node CI](https://github.com/andi1984/back2roots/workflows/Node%20CI/badge.svg)
[![Coverage Status](https://coveralls.io/repos/github/andi1984/back2roots/badge.svg?branch=develop)](https://coveralls.io/github/andi1984/back2roots?branch=develop)
![GitHub language count](https://img.shields.io/github/languages/count/andi1984/back2roots)
![GitHub code size in bytes](https://img.shields.io/github/languages/code-size/andi1984/back2roots)

Run `npm run start` to generate the HTML files based on the Markdown files saved inside the `md` folder.

The HTML content of the markdown files is going to be passed as `content` variable to the twig template specified in the respective `template` frontmatter of the markdown file. If none is specified, the `templates/main.twig` file is used as a fallback.

The output files are saved to the `dist` directory.

## NPM scripts

| Name          | Description                                                                              |
| ------------- | ---------------------------------------------------------------------------------------- |
| test          | Self-exploratory: Run your test suite here.                                              |
| tsc           | Runs the typescript compiler and checks for errors                                       |
| dev           | Starts a static file server and watches for content (e.g. template) or typescript files. |
| watch-content | Watches for content (template) changes                                                   |
| watch         | Watches for changes on the core typescript files for the static page generator itself.   |
| build         | Builds the output static HTML files.                                                     |
| assets        | Task for you to generate your assets with.                                               |
| start         | Build HTML files using `build` task and builds assets with `assets` task.                |
| release       | Command to generate a new release version.                                               |

## Twig Template Variables

| Variable |                     Description                      |                                                      Example |
| -------- | :--------------------------------------------------: | -----------------------------------------------------------: |
| title    |                  Title of the page                   |                                "Foo" (default: The filename) |
| sites    |               Array of all build sites               |                          cf. [Site interface](./utils/md.ts) |
| content  |             Content of the page as HTML              |                                 cf. `"<h1>Hello World</h1>"` |
| url      |             The URL of the current page.             |                    e.g. www.my-base-url.com/hello-world.html |
| base_url |               The base url of the page               |                                    cf. URL in env vars below |
| meta     | All metadata provided via frontmatter for given site | cf. {title: 'foo', template: 'bar', author: 'Max Mustermann' |

## CLI Options

| Option  | Alias |                                                                        Description | Default |
| ------- | :---: | ---------------------------------------------------------------------------------: | ------- |
| --clean |  --c  | Flag to indicate whether distribution folder should be cleaned before every build. | false   |

## Environment Variables

| Variable |                                                Description                                                 |                      Example |
| -------- | :--------------------------------------------------------------------------------------------------------: | ---------------------------: |
| URL      |                                              URL of your page                                              | `URL: https://www.google.de` |
| PRETTY   | indicate whether hosting can handle pretty URLs (e.g. `/hello/world` instead of `/hello/world/index.html`) |               `PRETTY: true` |

## Configuration

You can configure certain parts of the system. All config files are contained within the config folder.

### Hooks

| method       |                                Description                                 |                                  Example |
| ------------ | :------------------------------------------------------------------------: | ---------------------------------------: |
| sitesSorting | Sorts the sites before any further action (e.g. generating the HTML files) | Sort by date metadata/frontmatter field. |


CHANGELOG.md
# Changelog

## [v0.8.0](https://github.com/andi1984/back2roots/tree/v0.8.0) (2020-12-07)

[Full Changelog](https://github.com/andi1984/back2roots/compare/v0.7.0...v0.8.0)

**Fixed bugs:**

- Fix --pretty option to make it independent of Netlify [\#224](https://github.com/andi1984/back2roots/issues/224)

**Merged pull requests:**

- fix\(pretty-urls\): Limiting pretty URLs for trailing index URL segment [\#226](https://github.com/andi1984/back2roots/pull/226) ([andi1984](https://github.com/andi1984))
- build\(deps\): bump yargs from 16.1.0 to 16.2.0 [\#225](https://github.com/andi1984/back2roots/pull/225) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump @types/jest from 26.0.15 to 26.0.16 [\#223](https://github.com/andi1984/back2roots/pull/223) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump twig from 1.15.2 to 1.15.4 [\#222](https://github.com/andi1984/back2roots/pull/222) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps-dev\): bump release-it from 14.2.1 to 14.2.2 [\#221](https://github.com/andi1984/back2roots/pull/221) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump @types/node from 14.14.7 to 14.14.10 [\#220](https://github.com/andi1984/back2roots/pull/220) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump typescript from 4.0.5 to 4.1.2 [\#219](https://github.com/andi1984/back2roots/pull/219) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump @types/markdown-it from 10.0.2 to 10.0.3 [\#215](https://github.com/andi1984/back2roots/pull/215) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))

## [v0.7.0](https://github.com/andi1984/back2roots/tree/v0.7.0) (2020-11-10)

[Full Changelog](https://github.com/andi1984/back2roots/compare/v0.6.1...v0.7.0)

**Security fixes:**

- build\(deps\): \[security\] bump node-fetch from 2.6.0 to 2.6.1 [\#168](https://github.com/andi1984/back2roots/pull/168) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): \[security\] bump minimist from 1.2.0 to 1.2.5 [\#157](https://github.com/andi1984/back2roots/pull/157) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))

**Merged pull requests:**

- build\(deps\): bump @types/node from 14.6.2 to 14.14.7 [\#213](https://github.com/andi1984/back2roots/pull/213) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps-dev\): bump ts-jest from 26.3.0 to 26.4.4 [\#212](https://github.com/andi1984/back2roots/pull/212) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps-dev\): bump jest from 26.4.2 to 26.6.3 [\#211](https://github.com/andi1984/back2roots/pull/211) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps-dev\): bump release-it from 14.0.3 to 14.2.1 [\#210](https://github.com/andi1984/back2roots/pull/210) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump typescript from 4.0.3 to 4.0.5 [\#206](https://github.com/andi1984/back2roots/pull/206) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump markdown-it from 11.0.1 to 12.0.2 [\#202](https://github.com/andi1984/back2roots/pull/202) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump @types/jest from 26.0.14 to 26.0.15 [\#198](https://github.com/andi1984/back2roots/pull/198) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump ts-node-dev from 1.0.0-pre.63 to 1.0.0 [\#193](https://github.com/andi1984/back2roots/pull/193) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump yargs from 15.4.1 to 16.1.0 [\#192](https://github.com/andi1984/back2roots/pull/192) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump markdown-it-front-matter from 0.2.1 to 0.2.3 [\#189](https://github.com/andi1984/back2roots/pull/189) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump ts-node-dev from 1.0.0-pre.61 to 1.0.0-pre.63 [\#180](https://github.com/andi1984/back2roots/pull/180) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump typescript from 3.9.7 to 4.0.3 [\#178](https://github.com/andi1984/back2roots/pull/178) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump @types/jest from 26.0.10 to 26.0.14 [\#175](https://github.com/andi1984/back2roots/pull/175) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump markdown-it from 11.0.0 to 11.0.1 [\#174](https://github.com/andi1984/back2roots/pull/174) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps-dev\): bump release-it from 13.7.0 to 14.0.3 [\#173](https://github.com/andi1984/back2roots/pull/173) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- build\(deps\): bump twig from 1.15.1 to 1.15.2 [\#159](https://github.com/andi1984/back2roots/pull/159) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))

## [v0.6.1](https://github.com/andi1984/back2roots/tree/v0.6.1) (2020-09-01)

[Full Changelog](https://github.com/andi1984/back2roots/compare/v0.6.0...v0.6.1)

**Implemented enhancements:**

- Provide current page's URL into Twig as template variable [\#156](https://github.com/andi1984/back2roots/issues/156)

**Security fixes:**

- build\(deps\): \[security\] bump dot-prop from 4.2.0 to 4.2.1 [\#142](https://github.com/andi1984/back2roots/pull/142) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))

utils/__tests__/url.test.ts
const { isAbsoluteURL, generateURL } = require("../url");

jest.mock("../args");

describe("isAbsoluteURL", () => {
  test("it fails if url is not given", () => {
    expect(() => isAbsoluteURL()).toThrow();
  });

  test("it returns true if url is absolute", () => {
    expect(isAbsoluteURL("https://www.google.com")).toBe(true);
  });

  test("it returns false if url is relative", () => {
    expect(isAbsoluteURL("../foo")).toBe(false);
  });
});

describe("generateURL", () => {
  // Restore original env vars
  // NOTE: https://stackoverflow.com/a/48042799/778340
  const OLD_ENV = process.env;

  beforeEach(() => {
    jest.resetModules(); // most important - it clears the cache
    process.env = { ...OLD_ENV }; // make a copy
  });

  afterAll(() => {
    process.env = OLD_ENV; // restore old env
  });

  test("it fails if url is not given or not valid", () => {
    expect(() => generateURL()).toThrow();
  });

  test("pathname", () => {
    expect(generateURL("http://www.google.de").pathname).toBe("/");
    expect(generateURL("http://www.google.de/foo").pathname).toBe("/foo");
  });

  test("href", () => {
    expect(generateURL("http://www.google.de").href).toBe(
      "http://www.google.de/"
    );

    expect(generateURL("http://www.google.de/foo?myquery=2").href).toBe(
      "http://www.google.de/foo?myquery=2"
    );
  });

  test("Removes /index on pretty mode", () => {
    process.env.PRETTY = "true";
    process.env.URL = "http://foo.de";
    expect(generateURL("subfolder/index").href).toBe("http://foo.de/subfolder");
  });

  test("Does not remove /index.html on non-pretty mode", () => {
    process.env.PRETTY = "false";
    process.env.URL = "http://foo.de";
    expect(generateURL("subfolder/index").href).toBe(
      "http://foo.de/subfolder/index.html"
    );
  });

  test("Does not remove trailing segments other than index on pretty mode", () => {
    process.env.PRETTY = "true";
    process.env.URL = "http://foo.de";
    expect(generateURL("subfolder/impressum").href).toBe(
      "http://foo.de/subfolder/impressum.html"
    );
  });
});


utils/url.ts
import { URL } from "url";

// https://stackoverflow.com/a/41975448/778340
export {};

/**
 * Returns whether URL is absolute or not.
 * @param {string} url
 * @returns {boolean}
 */
const isAbsoluteURL = (url: string): Boolean => {
  return url.indexOf("://") > 0 || url.indexOf("//") === 0;
};

/**
 * Generates a new URL for the given link.
 * @param {string} link
 * @returns {URL}
 */
const generateURL = (link: string): URL => {
  if (isAbsoluteURL(link)) {
    return new URL(link);
  }

  if (process.env.PRETTY === "true" && link.endsWith("index")) {
    // In case we have support for pretty urls and link ends with index, we omit the last URL segment
    return new URL(link.replace(/\/index$/g, ""), process.env.URL);
  }

  return new URL(`${link}.html`, process.env.URL);
};

module.exports = {
  // https://stackoverflow.com/a/38979205/778340
  isAbsoluteURL,
  generateURL,
};


utils/renderMDFile.ts
// https://stackoverflow.com/a/41975448/778340
export {};

import { Site } from "./interfaces";

const path = require("path");
const fs = require("fs");

const minify = require("html-minifier").minify;
const Twig = require("twig");

const DISTRIBUTION_FOLDER = require("./constants").DISTRIBUTION_FOLDER;

module.exports = async (site: Site, sites: Site[]) => {
  if (!fs.existsSync(DISTRIBUTION_FOLDER)) {
    fs.mkdirSync(DISTRIBUTION_FOLDER);
  }

  // NOTE: Workaround of https://github.com/twigjs/twig.js/issues/509 to fail
  // early in case of template does not exists, because renderFile is failing
  // silently in that case.
  if (!fs.existsSync(site.template)) {
    throw new Error(`Template ${site.template} does not exist.`);
  }

  Twig.renderFile(
    site.template,
    {
      title: site.title,
      sites,
      content: site.html,
      meta: site.meta,
      url: site.url,
      base_url: process.env.URL,
    },
    (err: Error, html: string) => {
      return new Promise<void>((resolve, reject) => {
        if (err) {
          reject(err);
        }

        const outputFilePath = path.join(
          DISTRIBUTION_FOLDER,
          site.file.rel,
          `${site.file.name}.html`
        );

        // Create folder in case it doesn't exist already.
        if (!fs.existsSync(path.dirname(outputFilePath))) {
          fs.mkdirSync(path.dirname(outputFilePath), {
            recursive: true,
          });
        }

        // Write the file into the folder
        fs.writeFile(
          outputFilePath,
          minify(html, {
            collapseWhitespace: true,
          }),
          (writeErr: Error) => {
            if (writeErr) {
              reject(writeErr);
            } else {
              resolve();
            }
          }
        );
      });
    }
  );
};


build.ts
// https://stackoverflow.com/a/41975448/778340
export {};
import { Site } from "./utils/interfaces";
require("dotenv").config();

const args = require("./utils/args");
const path = require("path");
const fs = require("fs");
const { sync: syncRmRf } = require("rimraf");
const MARKDOWN_FOLDER = require("./utils/constants").MARKDOWN_FOLDER;
const DISTRIBUTION_FOLDER = require("./utils/constants").DISTRIBUTION_FOLDER;
const renderMarkdownFile = require("./utils/renderMDFile");
const { getFiles, filter } = require("./utils/generators");
const { getSiteObject } = require("./utils/md");
const { sitesSorting } = require("./config/hooks");
var RSS = require("rss");

module.exports = () => {
  // Step 1: Clear output directory
  if (args.clean) {
    syncRmRf(DISTRIBUTION_FOLDER);
  }

  // Step 2: Create files
  (async () => {
    let sites: Site[] = [];

    /* lets create an rss feed */
    var feed = new RSS({
      title: process.env.URL,
      description: `${process.env.URL} - RSS`,
      feed_url: `${process.env.URL}/rss.xml`,
      site_url: process.env.URL,
      image_url: `${process.env.URL}/icon.png`,
      // managingEditor: "Dylan Greene",
      // webMaster: "Dylan Greene",
      // copyright: "2013 Dylan Greene",
      language: "en",
      // categories: ["Category 1", "Category 2", "Category 3"],
      pubDate: new Date().toString(),
      ttl: "60",
    });

    // cache the xml to send to clients
    for await (const f of filter(
      getFiles(MARKDOWN_FOLDER),
      (filePath: string) => {
        const matchesMDFiletype = filePath.match(/^.*\.md$/g);
        return matchesMDFiletype && matchesMDFiletype.length > 0;
      }
    )) {
      sites.push(getSiteObject(f));
    }

    // Apply "sites" hooks
    sites.sort(sitesSorting);

    for (const site of sites) {
      renderMarkdownFile(site, sites);
      /* loop over data and add to feed */
      feed.item({
        title: site.title,
        description: site.html,
        url: site.url.href, // link to the item
        // author: "Guest Authors", // optional - defaults to feed author property
        // date: "May 27, 2012" // any format that js Date can parse.
      });
    }

    var xml = feed.xml();
    fs.writeFile(
      path.join(DISTRIBUTION_FOLDER, "rss.xml"),
      xml,
      (writeErr: Error) => {
        if (writeErr) {
          throw writeErr;
        }
      }
    );
  })();
};
